### PR TITLE
Add CGPO-style constrained reward utilities

### PR DESCRIPTION
**Pull request by githubshaurya**
## Summary

This PR adds CGPO-style constrained reward utilities that can be used inside OpenRLHF custom Python reward functions loaded through `--reward.remote_url`.

It does not add a new trainer or modify PPO/GRPO internals. Instead, it provides a small opt-in utility layer for calibrated rewards and binary constraint judging, inspired by CGPO / Mixture-of-Judges style RLHF.

## Motivation

CGPO proposes using calibrated rewards and a Mixture of Judges to reduce reward hacking and support constrained / multi-objective RLHF. OpenRLHF already supports custom reward functions that receive `queries`, `prompts`, and `labels`, so this PR adds reusable helpers for that existing path rather than changing the training stack.

Relevant references:

- CGPO paper: [https://arxiv.org/abs/2409.20370](https://arxiv.org/abs/2409.20370)
- TRL CGPO tracking issue: [https://github.com/huggingface/trl/issues/2156](https://github.com/huggingface/trl/issues/2156)
- TRL Mixture-of-Judges PR: [https://github.com/huggingface/trl/pull/2159](https://github.com/huggingface/trl/pull/2159)
- TRL judges docs: [https://huggingface.co/docs/trl/main/en/judges](https://huggingface.co/docs/trl/main/en/judges)

## Changes

- Add `openrlhf/utils/constrained_reward.py`
  - `BaseBinaryJudge`
  - `AllTrueJudge`
  - `calibrated_reward`
  - `shape_constrained_rewards`

- Add `tests/test_constrained_reward.py`
  - tests calibrated reward math
  - tests binary judgment validation
  - tests AllTrue / Mixture-of-Judges aggregation
  - tests constraint penalty shaping
  - tests gold/reference-completion based judging

- Add `examples/reward/constrained_reward_func.py`
  - example custom reward function using `labels` from `--data.label_key` as gold/reference completions
  - returns `rewards`, `scores`, and `extra_logs`

## Tests

Passed locally:

- `python -m py_compile openrlhf/utils/constrained_reward.py`
- `python -m py_compile examples/reward/constrained_reward_func.py`
- `python -m pytest -q tests/test_constrained_reward.py`

Smoke-tested the example reward function:

- `python -c "from examples.reward.constrained_reward_func import reward_func; out=reward_func(['What is 2+2? The answer is 4.'], ['What is 2+2?'], ['4']); print(out); assert isinstance(out['rewards'], float); assert out['scores']==1"`

Smoke output:

`{'rewards': 0.5, 'scores': 1, 'extra_logs': {'raw_reward': 1.0, 'baseline_reward': 1.0, 'calibrated_reward': 0.5, 'constraint_judgment': 1, 'constraint_penalty_applied': 0.0}}`

### CODE REVIEW

File: examples/reward/constrained_reward_func.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        judge=judge,
+        constraint_penalty=1.0,
+        calibrate=True,
+    )
+
+    return {
+        "rewards": result.rewards[0],
+        "scores": result.scores[0] if result.scores is not None else None,
+        "extra_logs": {key: values[0] for key, values in result.extra_logs.items()},
+    }
```
[the remark this answers is not in the document]
Comment (githubshaurya) [reply examples/reward/constrained_reward_func.py]: resolved

File: examples/reward/constrained_reward_func.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    spec.loader.exec_module(constrained_reward_module)
+
+    AllTrueJudge = constrained_reward_module.AllTrueJudge
+    BaseBinaryJudge = constrained_reward_module.BaseBinaryJudge
+    shape_constrained_rewards = constrained_reward_module.shape_constrained_rewards
+
+
+class NonEmptyJudge(BaseBinaryJudge):
+    def judge(self, prompts, completions, gold_completions=None):
+        return [int(bool(str(completion).strip())) for completion in completions]
```
[the remark this answers is not in the document]
Comment (githubshaurya) [reply examples/reward/constrained_reward_func.py]: resolved

File: examples/reward/constrained_reward_func.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        if gold_completions is None:
+            return [-1 for _ in completions]
+
+        judgments = []
+        for completion, gold in zip(completions, gold_completions):
+            completion = str(completion).strip().lower()
+            gold = str(gold).strip().lower()
+            judgments.append(int(bool(gold) and gold in completion))
+
+        return judgments
```
[the remark this answers is not in the document]
Comment (githubshaurya) [reply examples/reward/constrained_reward_func.py:37]: resolved

File: openrlhf/utils/constrained_reward.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        scores=scores,
+        extra_logs=extra_logs,
+    )
+
+
+def _as_float_tensor(values: Sequence[float] | torch.Tensor) -> torch.Tensor:
+    if isinstance(values, torch.Tensor):
+        return values.detach().float().cpu()
+    return torch.tensor(list(values), dtype=torch.float32)
+
```
[the remark this answers is not in the document]
Comment (githubshaurya) [reply openrlhf/utils/constrained_reward.py:180]: resolved

File: openrlhf/utils/constrained_reward.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    R_calib = sigmoid((R_generated - R_baseline) / temperature)
+    """
+
+    if temperature <= 0:
+        raise ValueError(f"temperature must be positive, got {temperature}")
+
+    rewards_tensor = _as_float_tensor(rewards)
+    baseline_tensor = _as_float_tensor(baseline_rewards)
+
```
[the remark this answers is not in the document]
Comment (githubshaurya) [reply openrlhf/utils/constrained_reward.py]: resolved
